File: backend/src/parser/enhanced_groff_parser.py
"""
Enhanced groff parser with comprehensive man page macro support.
"""

import re
from typing import Dict, List, Tuple, Optional, Any
from dataclasses import dataclass, field
from enum import Enum
import logging


logger = logging.getLogger(__name__)


class MacroType(Enum):
    """Man page macro types."""
    TH = "TH"  # Title header
    SH = "SH"  # Section header
    SS = "SS"  # Subsection header
    TP = "TP"  # Tagged paragraph
    PP = "PP"  # Plain paragraph
    LP = "LP"  # Left paragraph
    IP = "IP"  # Indented paragraph
    HP = "HP"  # Hanging paragraph
    RS = "RS"  # Relative indent start
    RE = "RE"  # Relative indent end
    BR = "BR"  # Bold/Roman alternating
    RB = "RB"  # Roman/Bold alternating
    IR = "IR"  # Italic/Roman alternating
    RI = "RI"  # Roman/Italic alternating
    BI = "BI"  # Bold/Italic alternating
    IB = "IB"  # Italic/Bold alternating
    B = "B"    # Bold
    I = "I"    # Italic
    R = "R"    # Roman
    SM = "SM"  # Small
    SB = "SB"  # Small bold
    P = "P"    # Paragraph
    OP = "OP"  # Option
    EE = "EE"  # Example end
    EX = "EX"  # Example start
    fi = "fi"  # Fill mode
    nf = "nf"  # No fill mode
    ad = "ad"  # Adjust
    na = "na"  # No adjust
    sp = "sp"  # Vertical space
    br = "br"  # Break
    bp = "bp"  # Break page


@dataclass
class ParsedSection:
    """Represents a parsed section of a man page."""
    name: str
    content: List[str] = field(default_factory=list)
    subsections: List['ParsedSection'] = field(default_factory=list)
    level: int = 1
    metadata: Dict[str, Any] = field(default_factory=dict)


@dataclass
class ManPageInfo:
    """Man page metadata from TH macro."""
    title: str = ""
    section: str = ""
    date: str = ""
    source: str = ""
    manual: str = ""


class EnhancedGroffParser:
    """
    Enhanced groff parser with full macro support.
    
    Features:
    - Complete man page macro handling
    - Proper text formatting and escaping
    - Section hierarchy preservation
    - Cross-reference detection
    - Code block identification
    """
    
    def __init__(self):
        """Initialize the enhanced parser."""
        self.reset()
        
    def reset(self):
        """Reset parser state."""
        self.info = ManPageInfo()
        self.sections: List[ParsedSection] = []
        self.current_section: Optional[ParsedSection] = None
        self.current_subsection: Optional[ParsedSection] = None
        self.indent_level = 0
        self.fill_mode = True
        self.in_example = False
        self.pending_tag = None
        self.font_stack = ['R']
        
    def parse(self, content: str) -> Dict[str, Any]:
        """
        Parse man page content with full macro support.
        
        Args:
            content: Raw man page content
            
        Returns:
            Parsed document structure
        """
        self.reset()
        lines = content.split('\n')
        
        i = 0
        while i < len(lines):
            line = lines[i].rstrip()
            
            # Skip empty lines in no-fill mode
            if not line and not self.fill_mode:
                self._add_content("")
                i += 1
                continue
            
            # Handle macro lines
            if line.startswith('.'):
                i = self._handle_macro(lines, i)
            else:
                # Handle text lines
                self._handle_text_line(line)
                i += 1
        
        # Finalize any pending content
        self._finalize_sections()
        
        return self._build_output()
    
    def _handle_macro(self, lines: List[str], index: int) -> int:
        """Handle macro commands."""
        line = lines[index]
        
        # Extract macro and arguments
        parts = line.split(None, 1)
        if not parts:
            return index + 1
        
        macro = parts[0][1:]  # Remove leading dot
        args = parts[1] if len(parts) > 1 else ""
        
        # Get macro handler
        handler = getattr(self, f'_handle_{macro}', None)
        if handler:
            return handler(lines, index, args)
        else:
            # Unknown macro - try generic handling
            return self._handle_generic_macro(lines, index, macro, args)
    
    def _handle_TH(self, lines: List[str], index: int, args: str) -> int:
        """Handle title header macro."""
        parts = self._parse_args(args)
        self.info.title = parts[0] if parts else ""
        self.info.section = parts[1] if len(parts) > 1 else ""
        self.info.date = parts[2] if len(parts) > 2 else ""
        self.info.source = parts[3] if len(parts) > 3 else ""
        self.info.manual = parts[4] if len(parts) > 4 else ""
        return index + 1
    
    def _handle_SH(self, lines: List[str], index: int, args: str) -> int:
        """Handle section header macro."""
        # Finalize previous section
        self._finalize_subsection()
        
        # Get section name
        if args:
            section_name = args.strip('"')
        else:
            # Next line contains section name
            index += 1
            if index < len(lines):
                section_name = lines[index].strip('"')
            else:
                section_name = "UNKNOWN"
        
        # Create new section
        self.current_section = ParsedSection(
            name=section_name.upper(),
            level=1
        )
        self.sections.append(self.current_section)
        self.current_subsection = None
        
        return index + 1
    
    def _handle_SS(self, lines: List[str], index: int, args: str) -> int:
        """Handle subsection header macro."""
        # Finalize previous subsection
        self._finalize_subsection()
        
        # Get subsection name
        if args:
            subsection_name = args.strip('"')
        else:
            # Next line contains subsection name
            index += 1
            if index < len(lines):
                subsection_name = lines[index].strip('"')
            else:
                subsection_name = "UNKNOWN"
        
        # Create new subsection
        if self.current_section:
            self.current_subsection = ParsedSection(
                name=subsection_name,
                level=2
            )
            self.current_section.subsections.append(self.current_subsection)
        
        return index + 1
    
    def _handle_TP(self, lines: List[str], index: int, args: str) -> int:
        """Handle tagged paragraph macro."""
        # Next line is the tag
        index += 1
        if index < len(lines):
            self.pending_tag = self._format_text(lines[index])
        return index + 1
    
    def _handle_IP(self, lines: List[str], index: int, args: str) -> int:
        """Handle indented paragraph macro."""
        tag = self._parse_args(args)[0] if args else "•"
        self._add_content(f"{tag} ", indent=True)
        return index + 1
    
    def _handle_PP(self, lines: List[str], index: int, args: str) -> int:
        """Handle plain paragraph macro."""
        self._add_content("")  # Add blank line
        return index + 1
    
    def _handle_BR(self, lines: List[str], index: int, args: str) -> int:
        """Handle bold/roman alternating macro."""
        parts = self._parse_args(args)
        formatted = []
        for i, part in enumerate(parts):
            if i % 2 == 0:
                formatted.append(f"**{part}**")
            else:
                formatted.append(part)
        self._add_content(" ".join(formatted))
        return index + 1
    
    def _handle_B(self, lines: List[str], index: int, args: str) -> int:
        """Handle bold macro."""
        if args:
            self._add_content(f"**{args}**")
        else:
            self.font_stack.append('B')
        return index + 1
    
    def _handle_I(self, lines: List[str], index: int, args: str) -> int:
        """Handle italic macro."""
        if args:
            self._add_content(f"*{args}*")
        else:
            self.font_stack.append('I')
        return index + 1
    
    def _handle_EX(self, lines: List[str], index: int, args: str) -> int:
        """Handle example start macro."""
        self.in_example = True
        self.fill_mode = False
        self._add_content("```")
        return index + 1
    
    def _handle_EE(self, lines: List[str], index: int, args: str) -> int:
        """Handle example end macro."""
        self.in_example = False
        self.fill_mode = True
        self._add_content("```")
        return index + 1
    
    def _handle_nf(self, lines: List[str], index: int, args: str) -> int:
        """Handle no-fill mode macro."""
        self.fill_mode = False
        if not self.in_example:
            self._add_content("```")
        return index + 1
    
    def _handle_fi(self, lines: List[str], index: int, args: str) -> int:
        """Handle fill mode macro."""
        if not self.fill_mode and not self.in_example:
            self._add_content("```")
        self.fill_mode = True
        return index + 1
    
    def _handle_RS(self, lines: List[str], index: int, args: str) -> int:
        """Handle relative indent start."""
        self.indent_level += 1
        return index + 1
    
    def _handle_RE(self, lines: List[str], index: int, args: str) -> int:
        """Handle relative indent end."""
        self.indent_level = max(0, self.indent_level - 1)
        return index + 1
    
    def _handle_sp(self, lines: List[str], index: int, args: str) -> int:
        """Handle vertical space."""
        self._add_content("")
        return index + 1
    
    def _handle_br(self, lines: List[str], index: int, args: str) -> int:
        """Handle line break."""
        self._add_content("  ")  # Markdown line break
        return index + 1
    
    def _handle_generic_macro(self, lines: List[str], index: int, macro: str, args: str) -> int:
        """Handle unknown macros generically."""
        # Log unknown macro for debugging
        logger.debug(f"Unknown macro: .{macro} {args}")
        return index + 1
    
    def _handle_text_line(self, line: str):
        """Handle regular text line."""
        if self.pending_tag:
            # This is the content for a tagged paragraph
            self._add_content(f"{self.pending_tag}\n    {self._format_text(line)}")
            self.pending_tag = None
        else:
            formatted = self._format_text(line)
            if formatted or not self.fill_mode:
                self._add_content(formatted)
    
    def _format_text(self, text: str) -> str:
        """Format text with escape sequences and font changes."""
        # Handle common escape sequences
        text = self._process_escapes(text)
        
        # Handle inline font changes
        text = self._process_inline_fonts(text)
        
        # Handle cross-references
        text = self._process_references(text)
        
        return text
    
    def _process_escapes(self, text: str) -> str:
        """Process groff escape sequences."""
        # Common escape sequences
        replacements = {
            r'\-': '-',
            r'\ ': ' ',
            r'\e': '\\',
            r'\&': '',
            r'\~': ' ',
            r'\(aq': "'",
            r'\(cq': "'",
            r'\(oq': "'",
            r'\(dq': '"',
            r'\(em': '—',
            r'\(en': '–',
            r'\(bu': '•',
            r'\(co': '©',
            r'\(rg': '®',
            r'\(tm': '™',
            r'\*(': '',  # String variable (simplified)
            r'\fB': '**',  # Bold start
            r'\fI': '*',   # Italic start
            r'\fR': '',    # Roman (reset)
            r'\fP': '',    # Previous font
        }
        
        for escape, replacement in replacements.items():
            text = text.replace(escape, replacement)
        
        # Handle \f[BIPR] font changes
        text = re.sub(r'\\f\[([BIPR])\]', self._font_to_markdown, text)
        
        # Remove other escape sequences
        text = re.sub(r'\\[a-zA-Z]\(..', '', text)
        text = re.sub(r'\\[a-zA-Z].', '', text)
        text = re.sub(r'\\[a-zA-Z]', '', text)
        
        return text
    
    def _font_to_markdown(self, match):
        """Convert font codes to markdown."""
        font = match.group(1)
        if font == 'B':
            return '**'
        elif font == 'I':
            return '*'
        else:
            return ''
    
    def _process_inline_fonts(self, text: str) -> str:
        """Process inline font changes."""
        # Handle \fBtext\fR patterns
        text = re.sub(r'\\fB([^\\]+)\\fR', r'**\1**', text)
        text = re.sub(r'\\fI([^\\]+)\\fR', r'*\1*', text)
        
        # Handle alternating fonts
        text = re.sub(r'\\fB([^\\]+)\\fI([^\\]+)\\fR', r'**\1***\2*', text)
        
        return text
    
    def _process_references(self, text: str) -> str:
        """Process man page cross-references."""
        # Match patterns like command(section)
        def make_ref(match):
            cmd = match.group(1)
            section = match.group(2)
            return f"[{cmd}({section})]({cmd})"
        
        text = re.sub(r'\b([a-zA-Z][\w\-\.]*)\((\d+)\)', make_ref, text)
        return text
    
    def _parse_args(self, args: str) -> List[str]:
        """Parse macro arguments respecting quotes."""
        if not args:
            return []
        
        # Simple quote-aware splitting
        parts = []
        current = []
        in_quotes = False
        
        for char in args:
            if char == '"' and not in_quotes:
                in_quotes = True
            elif char == '"' and in_quotes:
                in_quotes = False
                parts.append(''.join(current))
                current = []
            elif char == ' ' and not in_quotes:
                if current:
                    parts.append(''.join(current))
                    current = []
            else:
                current.append(char)
        
        if current:
            parts.append(''.join(current))
        
        return parts
    
    def _add_content(self, text: str, indent: bool = False):
        """Add content to current section."""
        if indent and self.indent_level > 0:
            text = "    " * self.indent_level + text
        
        if self.current_subsection:
            self.current_subsection.content.append(text)
        elif self.current_section:
            self.current_section.content.append(text)
    
    def _finalize_subsection(self):
        """Finalize current subsection."""
        # Clean up trailing empty lines
        if self.current_subsection and self.current_subsection.content:
            while self.current_subsection.content and not self.current_subsection.content[-1]:
                self.current_subsection.content.pop()
    
    def _finalize_sections(self):
        """Finalize all sections."""
        self._finalize_subsection()
        
        # Clean up section content
        for section in self.sections:
            if section.content:
                while section.content and not section.content[-1]:
                    section.content.pop()
    
    def _build_output(self) -> Dict[str, Any]:
        """Build final output structure."""
        return {
            'title': self.info.title,
            'section': self.info.section,
            'date': self.info.date,
            'source': self.info.source,
            'manual': self.info.manual,
            'sections': [self._section_to_dict(s) for s in self.sections]
        }
    
    def _section_to_dict(self, section: ParsedSection) -> Dict[str, Any]:
        """Convert section to dictionary."""
        return {
            'name': section.name,
            'content': '\n'.join(section.content),
            'subsections': [self._section_to_dict(s) for s in section.subsections]
        }

File: frontend/src/utils/enhancedGroffParser.ts
/**
 * Enhanced Groff parser for better document rendering
 * Handles various groff/troff formatting with improved accuracy
 */

export interface ParsedSection {
  name: string;
  content: string;
  subsections?: ParsedSubsection[];
  type?: string;
  metadata?: Record<string, any>;
}

export interface ParsedSubsection {
  name: string;
  content: string;
  type?: string;
}

export interface ParseOptions {
  preserveFormatting?: boolean;
  detectCodeBlocks?: boolean;
  enhanceLinks?: boolean;
  parseOptions?: boolean;
}

export class EnhancedGroffParser {
  private readonly macroMap: Map<string, (args: string[]) => string> = new Map();
  
  constructor() {
    this.initializeMacros();
  }

  private initializeMacros(): void {
    // Common groff macros
    this.macroMap.set('.TH', () => ''); // Title header - handled separately
    this.macroMap.set('.SH', (args) => `\n## ${args.join(' ')}\n`); // Section header
    this.macroMap.set('.SS', (args) => `\n### ${args.join(' ')}\n`); // Subsection
    this.macroMap.set('.PP', () => '\n\n'); // Paragraph
    this.macroMap.set('.P', () => '\n\n'); // Paragraph
    this.macroMap.set('.LP', () => '\n\n'); // Left paragraph
    this.macroMap.set('.TP', () => '\n'); // Tagged paragraph
    this.macroMap.set('.IP', (args) => `\n• ${args[0] || ''}`); // Indented paragraph
    this.macroMap.set('.HP', () => '\n'); // Hanging paragraph
    this.macroMap.set('.RS', () => ''); // Start relative indent
    this.macroMap.set('.RE', () => ''); // End relative indent
    this.macroMap.set('.B', (args) => `**${args.join(' ')}**`); // Bold
    this.macroMap.set('.I', (args) => `*${args.join(' ')}*`); // Italic
    this.macroMap.set('.BI', (args) => this.alternateFormat(args, true)); // Bold-italic
    this.macroMap.set('.BR', (args) => this.alternateFormat(args, false)); // Bold-roman
    this.macroMap.set('.IR', (args) => this.alternateItalicRoman(args)); // Italic-roman
    this.macroMap.set('.RB', (args) => this.reverseAlternateFormat(args)); // Roman-bold
    this.macroMap.set('.SM', (args) => args.join(' ')); // Small
    this.macroMap.set('.SB', (args) => `**${args.join(' ')}**`); // Small bold
    this.macroMap.set('.nf', () => '```\n'); // No fill (code block start)
    this.macroMap.set('.fi', () => '\n```'); // Fill (code block end)
    this.macroMap.set('.EX', () => '```\n'); // Example start
    this.macroMap.set('.EE', () => '\n```'); // Example end
    this.macroMap.set('.in', () => ''); // Indent
    this.macroMap.set('.sp', () => '\n'); // Space
    this.macroMap.set('.br', () => '  \n'); // Break
    this.macroMap.set('.ad', () => ''); // Adjust
    this.macroMap.set('.na', () => ''); // No adjust
    this.macroMap.set('.nh', () => ''); // No hyphenation
    this.macroMap.set('.hy', () => ''); // Hyphenation
  }

  private alternateFormat(args: string[], startBold: boolean): string {
    return args.map((arg, i) => {
      if (i % 2 === (startBold ? 0 : 1)) {
        return `**${arg}**`;
      }
      return arg;
    }).join(' ');
  }

  private alternateItalicRoman(args: string[]): string {
    return args.map((arg, i) => {
      if (i % 2 === 0) {
        return `*${arg}*`;
      }
      return arg;
    }).join(' ');
  }

  private reverseAlternateFormat(args: string[]): string {
    return args.map((arg, i) => {
      if (i % 2 === 1) {
        return `**${arg}**`;
      }
      return arg;
    }).join(' ');
  }

  public parseGroffContent(content: string, options: ParseOptions = {}): string {
    let result = content;
    
    // Remove backslash escapes for common characters
    result = result.replace(/\\-/g, '-');
    result = result.replace(/\\\s/g, ' ');
    result = result.replace(/\\\\/g, '\\');
    result = result.replace(/\\e/g, '\\');
    result = result.replace(/\\&/g, '');
    result = result.replace(/\\~/g, ' ');
    result = result.replace(/\\0/g, ' ');
    result = result.replace(/\\|/g, '');
    result = result.replace(/\\\^/g, '');
    
    // Process font changes inline
    result = result.replace(/\\fB([^\\]+)\\fR/g, '**$1**'); // Bold
    result = result.replace(/\\fI([^\\]+)\\fR/g, '*$1*'); // Italic
    result = result.replace(/\\fB([^\\]+)\\fP/g, '**$1**'); // Bold with previous font
    result = result.replace(/\\fI([^\\]+)\\fP/g, '*$1*'); // Italic with previous font
    
    // Process special characters
    result = result.replace(/\\'/g, '\'');
    result = result.replace(/\\`/g, '`');
    result = result.replace(/\\"/g, '"');
    result = result.replace(/\\\./g, '.');
    
    // Handle .TH (title) macros specially
    result = result.replace(/^\.TH\s+(.+)$/gm, (match, args) => {
      const parts = this.parseQuotedArgs(args);
      return `# ${parts[0] || ''} (${parts[1] || ''})`;
    });
    
    // Process all other macros
    const lines = result.split('\n');
    const processedLines: string[] = [];
    let inCodeBlock = false;
    let inExample = false;
    
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const trimmedLine = line.trim();
      
      // Handle code block markers
      if (trimmedLine === '.nf' || trimmedLine === '.EX') {
        inCodeBlock = true;
        processedLines.push('```');
        continue;
      }
      
      if (trimmedLine === '.fi' || trimmedLine === '.EE') {
        inCodeBlock = false;
        processedLines.push('```');
        continue;
      }
      
      // Skip certain macros in code blocks
      if (inCodeBlock && !trimmedLine.startsWith('.')) {
        processedLines.push(line);
        continue;
      }
      
      // Process macro lines
      if (trimmedLine.startsWith('.')) {
        const processed = this.processMacroLine(trimmedLine);
        if (processed !== null) {
          processedLines.push(processed);
        }
      } else {
        processedLines.push(line);
      }
    }
    
    result = processedLines.join('\n');
    
    // Clean up extra newlines
    result = result.replace(/\n{3,}/g, '\n\n');
    
    // Enhance links if requested
    if (options.enhanceLinks) {
      result = this.enhanceLinks(result);
    }
    
    // Auto-detect code blocks if requested
    if (options.detectCodeBlocks) {
      result = this.autoDetectCodeBlocks(result);
    }
    
    return result.trim();
  }

  private processMacroLine(line: string): string | null {
    const match = line.match(/^\.(\w+)(?:\s+(.*))?$/);
    if (!match) return line;
    
    const [, macro, args] = match;
    const handler = this.macroMap.get(`.${macro}`);
    
    if (handler) {
      const parsedArgs = args ? this.parseQuotedArgs(args) : [];
      return handler(parsedArgs);
    }
    
    // Unknown macro - return as comment
    return `<!-- ${line} -->`;
  }

  private parseQuotedArgs(argsString: string): string[] {
    const args: string[] = [];
    let current = '';
    let inQuotes = false;
    let quoteChar = '';
    
    for (let i = 0; i < argsString.length; i++) {
      const char = argsString[i];
      
      if (inQuotes) {
        if (char === quoteChar) {
          inQuotes = false;
          args.push(current);
          current = '';
        } else {
          current += char;
        }
      } else {
        if (char === '"' || char === "'") {
          inQuotes = true;
          quoteChar = char;
        } else if (char === ' ' || char === '\t') {
          if (current) {
            args.push(current);
            current = '';
          }
        } else {
          current += char;
        }
      }
    }
    
    if (current) {
      args.push(current);
    }
    
    return args;
  }

  private enhanceLinks(content: string): string {
    // Enhance man page references like command(section)
    content = content.replace(/\b(\w+)\((\d+)\)/g, (match, cmd, section) => {
      return `[${cmd}(${section})](/docs/${cmd}.${section})`;
    });
    
    // Enhance URLs
    content = content.replace(
      /(https?:\/\/[^\s<>"{}|\\^`[\]]+)/g,
      '[$1]($1)'
    );
    
    return content;
  }

  private autoDetectCodeBlocks(content: string): string {
    const lines = content.split('\n');
    const result: string[] = [];
    let possibleCodeBlock: string[] = [];
    let inDetectedCodeBlock = false;
    
    for (const line of lines) {
      // Heuristics for code detection
      const looksLikeCode = 
        /^\s{4,}/.test(line) || // Indented 4+ spaces
        /^\$\s/.test(line.trim()) || // Shell prompt
        /^#\s/.test(line.trim()) || // Root shell prompt
        /^[a-z_]+\s*\(.*\)\s*{/.test(line) || // Function definition
        /^[A-Z_]+=[^=]+$/.test(line.trim()); // Environment variable
      
      if (looksLikeCode && !inDetectedCodeBlock) {
        if (possibleCodeBlock.length > 0) {
          result.push(...possibleCodeBlock);
          possibleCodeBlock = [];
        }
        inDetectedCodeBlock = true;
        result.push('```bash');
        result.push(line);
      } else if (inDetectedCodeBlock && !looksLikeCode && line.trim() === '') {
        // Empty line in code block - continue
        result.push(line);
      } else if (inDetectedCodeBlock && !looksLikeCode) {
        // End of code block
        result.push('```');
        result.push(line);
        inDetectedCodeBlock = false;
      } else {
        result.push(line);
      }
    }
    
    if (inDetectedCodeBlock) {
      result.push('```');
    }
    
    return result.join('\n');
  }

  public parseGroffSections(
    sections: Array<{ name: string; content: string; subsections?: any[] },
    options: ParseOptions = {}
  ): ParsedSection[] {
    return sections.map(section => {
      const parsed: ParsedSection = {
        name: section.name,
        content: this.parseGroffContent(section.content, options),
        type: this.detectSectionType(section.name),
      };

      if (section.subsections && section.subsections.length > 0) {
        parsed.subsections = section.subsections.map(sub => ({
          name: sub.name,
          content: this.parseGroffContent(sub.content, options),
          type: 'subsection',
        }));
      }

      // Add metadata for specific section types
      if (parsed.type === 'options' && options.parseOptions) {
        parsed.metadata = {
          options: this.extractOptions(parsed.content),
        };
      }

      return parsed;
    });
  }

  private detectSectionType(name: string): string {
    const normalized = name.toUpperCase();
    if (/^NAME/.test(normalized)) return 'name';
    if (/^SYNOP/.test(normalized)) return 'synopsis';
    if (/^DESCR/.test(normalized)) return 'description';
    if (/^OPTION/.test(normalized)) return 'options';
    if (/^EXAMPLE/.test(normalized)) return 'examples';
    if (/^EXIT/.test(normalized)) return 'exit-status';
    if (/^ENVIRON/.test(normalized)) return 'environment';
    if (/^FILE/.test(normalized)) return 'files';
    if (/^SEE/.test(normalized)) return 'see-also';
    if (/^BUG/.test(normalized)) return 'bugs';
    if (/^AUTHOR/.test(normalized)) return 'author';
    if (/^COPYRIGHT|LICENSE/.test(normalized)) return 'copyright';
    return 'general';
  }

  private extractOptions(content: string): Array<{
    flags: string[];
    description: string;
    argument?: string;
  } {
    const options: Array<{ flags: string[]; description: string; argument?: string } = [];
    const lines = content.split('\n');
    
    let currentOption: { flags: string[]; description: string; argument?: string } | null = null;
    let descLines: string[] = [];
    
    for (const line of lines) {
      // Match option patterns like: -h, --help
      const optionMatch = line.match(/^\s*((?:-[\w-]+(?:\s*,\s*)?)+)(?:\s+(\S+))?\s*(.*)$/);
      
      if (optionMatch) {
        // Save previous option
        if (currentOption) {
          currentOption.description = descLines.join(' ').trim();
          options.push(currentOption);
          descLines = [];
        }
        
        // Parse new option
        const [, flagsStr, arg, desc] = optionMatch;
        const flags = flagsStr.split(/\s*,\s*/).map(f => f.trim());
        
        currentOption = {
          flags,
          description: '',
          ...(arg && { argument: arg }),
        };
        
        if (desc) descLines.push(desc);
      } else if (currentOption && line.trim()) {
        // Continue description
        descLines.push(line.trim());
      }
    }
    
    // Save last option
    if (currentOption) {
      currentOption.description = descLines.join(' ').trim();
      options.push(currentOption);
    }
    
    return options;
  }
}

// Export singleton instance
export const enhancedGroffParser = new EnhancedGroffParser();